proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,142 @@
+
+ * `Experimental` extensions are undergoing active development, or have consequences that make it impossible in practice to avoid breaking changes. The syntax and semantics that are enabled by the extension are likely to change regularly.
+   * It is expected that most new language extensions will begin as experimental, while the developers and the community gain experience with their use.
+   * Despite being open to breaking change, an `Experimental` extension must be `Deprecated` prior to removal.
+   * Some extensions in this category might be `RequiredTypeArguments`, `RebindableSyntax`, and `Arrows`.
+
+ * `Deprecated` extensions are considered to be design or implementation dead ends, and should not be used in new code. Deprecating an extension means that GHC may remove support for it in a future release (though this decision may be revised in the light of user feedback).
+   * When an extension is deprecated, the user's guide and the compiler warning _must_ include a statement about the longevity of the extension, though this need not necessarily commit to a concrete time for removal. Additionally, the user's guide and the warning should direct users to information about migrating away from the deprecated extension.
+   * Any extension will be deprecated prior to removal.
+   * Some extensions in this category might be `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
+
+ * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language edition, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Extensions that are `Legacy` may have some conflict with `Stable` extensions and explicitly do not need to be considered for an extension to be `Stable`. Some extensions in this category might be `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.
+
+Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.
+
+The following transitions are included:
+ * (does not exist) -> `Experimental`
+ * `Experimental` -> `Stable`
+ * `Experimental` -> `Deprecated`
+ * `Experimental` -> `Legacy`
+ * `Stable` -> `Deprecated`
+ * `Stable` -> `Legacy`
+ * `Legacy` -> `Deprecated`
+ * `Deprecated` -> (does not exist)
+
+These transitions are explicitly excluded:
+ * (does not exist) -> `Deprecated`
+ * (does not exist) -> `Legacy`
```

Comment:
I'm not convinced about excluding these. Something we have been known to do in the past is to introduce a new extension name for existing functionality (of the base language or another extension) that is being deprecated or otherwise restricted. For example:
 * `DeepSubsumption` was arguably `Legacy` when it was introduced, because its functionality was already present in previous releases but we wanted to move towards a world in which `NoDeepSubsumption` became the default. (I'm not sure whether we would have considered `NoDeepSubsumption` to be `Experimental` or `Stable` on its introduction.)
 * If we want to partially deprecate an extension, I think the right way is to split it into multiple extensions and introduce one or more of them as `Deprecated` (e.g. I can imagine splitting up `IncoherentInstances = IncoherentInstanceDeclarations + IncoherentClassRoles + IncoherentSolving` with `IncoherentInstanceDeclarations` being `Deprecated`).
 
So I'd suggest moving these to the "discretionary" category.

Reply:
> I'm not convinced about excluding these. Something we have been known to do in the past is to introduce a new extension name for existing functionality (of the base language or another extension) that is being deprecated or otherwise restricted. For example:
> 
> * `DeepSubsumption` was arguably `Legacy` when it was introduced, because its functionality was already present in previous releases but we wanted to move towards a world in which `NoDeepSubsumption` became the default. (I'm not sure whether we would have considered `NoDeepSubsumption` to be `Experimental` or `Stable` on its introduction.)

This is a really good point.

> * If we want to partially deprecate an extension, I think the right way is to split it into multiple extensions and introduce one or more of them as `Deprecated` (e.g. I can imagine splitting up `IncoherentInstances = IncoherentInstanceDeclarations + IncoherentClassRoles + IncoherentSolving` with `IncoherentInstanceDeclarations` being `Deprecated`).

I would have expected all of them to be `Experimental` at first. That way if the way `IncoherentInstances` was cut needed to change then it is less problematic as the entire set was `Experimental`. To be more concrete, if it was realized after being "in the wild" that it is better defined as `IncoherentInstances = IncoherentDeclarationsAndClassRoles + IncoherentSolving` or maybe `IncoherentInstances = IncoherentInstanceDeclarations + IncoherentClassRolesAndSolving` then the only removals/changes would have been made to `Experimental` extensions.

That said, I'm fine with moving this to the "discretionary" category.